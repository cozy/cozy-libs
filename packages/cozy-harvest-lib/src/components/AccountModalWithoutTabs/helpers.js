import get from 'lodash/get'

export const getMatchingTrigger = (accountsAndTriggers, accountId) => {
  return get(
    accountsAndTriggers.find(
      accountAndTrigger => accountAndTrigger.account._id === accountId
    ),
    'trigger'
  )
}
