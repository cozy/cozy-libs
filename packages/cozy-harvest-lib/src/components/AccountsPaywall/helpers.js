// @ts-check
// @ts-ignore
import flag from 'cozy-flags'

/**
 * @param {string} slug the slug of the konnector to check
 * @returns {number} the maximum accounts allowed for given konnector
 */
export function computeMaxAccountsByKonnector(slug) {
  const maxAccountsByKonnector = flag(`harvest.accounts.maxByKonnector.${slug}`)
  if (maxAccountsByKonnector !== null) {
    return maxAccountsByKonnector === -1 ? Infinity : maxAccountsByKonnector
  }
  const maxAccountsDefault = flag('harvest.accounts.maxByKonnector.default')
  if (maxAccountsDefault !== null) {
    return maxAccountsDefault === -1 ? Infinity : maxAccountsDefault
  }
  return Infinity
}

/**
 *
 * @param {string} slug the slug of the konnector to check
 * @param {number} nbAccounts the current number of account for the konnector to check
 * @returns {boolean} whether the number of accounts allowed for a given connector has been reached
 */
export function hasReachMaxAccountsByKonnector(slug, nbAccounts) {
  const maxAccounts = computeMaxAccountsByKonnector(slug)
  if (isFinite(maxAccounts) && nbAccounts >= maxAccounts) {
    return true
  }

  return false
}

/**
 * @returns {number} the maximum accounts allowed for all konnectors
 */
export function computeMaxAccounts() {
  const maxAccounts = flag('harvest.accounts.max')
  if (maxAccounts !== null) {
    return maxAccounts === -1 ? Infinity : maxAccounts
  }
  return Infinity
}

/**
 * @typedef {Object} KonnectorOffer
 * @property {string} slug - Slug of the konnector
 * @property {number} credit - Number of account credits for the konnector
 * @property {string} startsAt - Offer start date
 * @property {string} [endsAt] - Offer end date
 */

/**
 * Compute the number of account credits for all konnectors
 * @returns {number} - Number of account credits for all konnectors
 */
export function computeGeneralOfferCredits() {
  /** @type {KonnectorOffer[]} */
  const offers = flag('harvest.accounts.offers.list') || []
  const generalOffers = offers.filter(offer => offer.slug === '*')

  return generalOffers.reduce((total, offer) => {
    const now = new Date()
    const startsAt = new Date(offer.startsAt)
    const endsAt = offer.endsAt ? new Date(offer.endsAt) : new Date()
    if (now >= startsAt && now <= endsAt) {
      return total + offer.credit
    }

    return total
  }, 0)
}

/**
 *
 * @param {number} nbAccounts the current number of account for all konnectors
 * @returns {boolean} whether the number of accounts allowed for all konnectors has been reached
 */
export function hasReachMaxAccounts(nbAccounts) {
  const generalOffer = computeGeneralOfferCredits()
  const maxAccounts = computeMaxAccounts()

  if (isFinite(maxAccounts) && nbAccounts >= maxAccounts + generalOffer) {
    return true
  }

  return false
}

/**
 * Compute the number of account credits for a konnector
 * @param {string} slug - Slug of the konnector
 * @param {import('cozy-client/types/types').IOCozyTrigger[]} accounts - List of accounts of the konnector
 * @returns
 */
export const computeRemainingOfferCreditsByKonnector = (
  slug,
  accounts = []
) => {
  /** @type {KonnectorOffer[]} */
  const offers = flag('harvest.accounts.offers.list') || []
  const offersForKonnector = offers.filter(offer => offer.slug === slug)

  if (offersForKonnector.length === 0) {
    return 0
  }

  let listOfAccounts = accounts.sort(
    (a, b) =>
      new Date(a.cozyMetadata.createdAt).getTime() -
      new Date(b.cozyMetadata.createdAt).getTime()
  )

  return offersForKonnector.reduce((acc, offer) => {
    const startsAt = new Date(offer.startsAt).getTime()
    const endsAt = offer.endsAt ? new Date(offer.endsAt).getTime() : undefined

    if (
      (endsAt !== undefined && endsAt < Date.now()) ||
      startsAt > Date.now()
    ) {
      return acc
    }

    let accountCreditAvailable = offer.credit || 0
    listOfAccounts = listOfAccounts.filter(account => {
      const accountDate = new Date(account.cozyMetadata.createdAt).getTime()

      if (
        accountCreditAvailable > 0 &&
        accountDate >= startsAt &&
        (!endsAt || endsAt >= accountDate)
      ) {
        accountCreditAvailable -= 1
        return false
      }
      return true
    })

    return acc + accountCreditAvailable
  }, 0)
}
