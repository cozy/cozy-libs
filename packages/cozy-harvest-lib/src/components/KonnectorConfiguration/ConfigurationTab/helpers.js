import set from 'lodash/set'

import {
  setFields as setBankAccountFields,
  getAccountLabel
} from './bankAccountHelpers'

const setFieldsPerDoctype = {
  'io.cozy.bank.accounts': setBankAccountFields,
  default: set
}

export const updateContract = async (client, contract, options) => {
  const { fields, onSuccess, onError } = options

  const setFields =
    setFieldsPerDoctype[contract._type] || setFieldsPerDoctype.default
  setFields(contract, fields)

  try {
    await client.save(contract)
    onSuccess()
  } catch (err) {
    onError(err)
  }
}

const getPrimaryTextPerDoctype = {
  'io.cozy.bank.accounts': getAccountLabel
}

const getPrimaryTextDefault = contract => contract.label

export const getPrimaryText = contract => {
  const fn = getPrimaryTextPerDoctype[contract._type] || getPrimaryTextDefault
  return fn(contract)
}

export const isDisabled = contract => Boolean(contract?.metadata?.disabledAt)

export const isDeleted = contract => contract._deleted

export const isErrored = contract => Boolean(contract?.metadata?.error)

export const isImported = contract => Boolean(contract?.metadata?.imported)
