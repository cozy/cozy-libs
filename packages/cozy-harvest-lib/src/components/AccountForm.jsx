import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
