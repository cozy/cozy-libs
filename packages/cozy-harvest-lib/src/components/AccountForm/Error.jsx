import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { translate } from 'cozy-ui/react/I18n'

import { KonnectorJobError } from '../../helpers/konnectors'
import Markdown from '../Markdown'

/**
 * Return the locale key corresponding to error description for a given error
 * code.
 * @param  {string} code Error code as returned by a konnector job
 * @return {string}      locale key
 */
const getKonnectorJobKey = code => `error.job.${code}.description`

/**
 * Component used to show an error related to an account or to a konnector job
 * Regular error are treated in a common way, but for KonnectorJobErrors
 * we can add a little bit of logic.
 *
 * This component is strongly related with the content of locales files and
 * deals mainly with translation concerns.
 */
export class AccountFormError extends PureComponent {
  defaultKey = 'error.job.UNKNOWN_ERROR.description'

  /**
   * Returns the description of an error.
   * This method first try to get the full error code
   * (for example LOGIN_FAILED.LOGIN_FAILED.NEEDS_SECRET), then fallback on the
   * error type, which is the error code first segment (in our example, it's
   * LOGIN_FAILED). It none of these two tries returns anything, it means
   * that the error is unknown or not yet handled by harvest, so we fallback
   * to the default error messages.
   * @param  {Error} error      The error
   * @param  {Object} konnector konnector related to this error
   * @param  {Func} t           Translation function, expected to be Polyglot.t()
   * @return {String}           The error description
   */
  getDescription() {
    const { error, konnector, t } = this.props

    const args = {
      name: konnector.name || '',
      link: konnector.vendor_link || ''
    }

    if (error instanceof KonnectorJobError) {
      return t(getKonnectorJobKey(error.code), {
        ...args,
        _: t(getKonnectorJobKey(error.type), {
          ...args,
          _: t(this.defaultKey, args)
        })
      })
    }

    const description = t(this.defaultKey, args)
    return error.message ? `${description} (${error.message})` : description
  }

  render() {
    const { error } = this.props
    /* eslint-disable-next-line no-console */
    console.error(error)
    const errorDescription = this.getDescription()
    return (
      <div className="u-error">
        <Markdown source={errorDescription} />
      </div>
    )
  }
}

AccountFormError.proptTypes = {
  error: PropTypes.object.isRequired,
  konnector: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(AccountFormError)
