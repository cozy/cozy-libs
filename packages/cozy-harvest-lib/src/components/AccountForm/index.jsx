import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'

import AccountFields from './AccountFields'
import AccountFormError from './Error'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from '../../helpers/konnectors'
import manifest from '../../helpers/manifest'
import OAuthForm from '../../components/OAuthForm'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (showError || isLoginError) && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  showError: PropTypes.bool,
  submitting: PropTypes.bool
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(AccountForm)
