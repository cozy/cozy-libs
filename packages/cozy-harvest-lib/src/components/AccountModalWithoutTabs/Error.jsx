import React from 'react'

import { useClient } from 'cozy-client'
import Button from 'cozy-ui/transpiled/react/Buttons'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import Infos from 'cozy-ui/transpiled/react/Infos'

import { loadSelectedAccountId } from '../../helpers/accounts'
import withLocales from '../hoc/withLocales'

const Error = ({ accountId, accountsAndTriggers, trigger, lastError }) => {
  const { t } = useI18n()
  const client = useClient()

  const error = !trigger ? new Error('No matching trigger found') : lastError

  const handleClick = () => {
    loadSelectedAccountId(client, accountId, accountsAndTriggers)
  }

  return (
    <Infos
      actionButton={
        <Button
          label={t('modal.konnector.error.button')}
          color="error"
          onClick={handleClick}
        />
      }
      title={t('modal.konnector.error.title')}
      text={t('modal.konnector.error.description', error)}
      icon="warning"
      isImportant
    />
  )
}

export default withLocales(Error)
