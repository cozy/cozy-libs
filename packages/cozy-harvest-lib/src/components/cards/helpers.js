// @ts-check
import { formatLocallyDistanceToNow } from 'cozy-ui/transpiled/react/I18n/format'

import { isDisconnected } from '../../helpers/konnectors'
import { getLastSuccessDate } from '../../helpers/triggers'

const DEFAULT_TIME = 300_000 // To milliseconds (5 minutes)
/**
 *
 * @param {Date} date
 * @returns number
 */
const getDifferenceInMillisecondes = date => {
  return Date.now() - new Date(date).getTime()
}

/**
 * @param {object} options
 * @param {function} options.t - i18n function
 * @param {import('cozy-client/types/types').IOCozyKonnector} options.konnector - Associated Connector
 * @param {object} options.trigger - Associated trigger
 * @param {object} options.running - If the connector is running
 * @param {object} options.expectingTriggerLaunch - If the trigger is waiting to be launched
 * @param {object} options.lastSuccessDate - The last date when the trigger was successfully executed
 * @param {object} options.isKonnectorRunnable - If the konnector is runnable
 * @param {object} options.isInMaintenance - If the konnector is in maintenance
 * @returns {string}
 */
export const makeLabel = ({
  t,
  konnector,
  trigger,
  running,
  expectingTriggerLaunch,
  isInMaintenance,
  isKonnectorRunnable
}) => {
  const lastSuccessDate = getLastSuccessDate(trigger)
  if (!isKonnectorRunnable) {
    return t('accountForm.notClientSide', { name: konnector.name })
  }
  const mantenanceSuffix = isInMaintenance
    ? ` Â· ${t('konnectorBlock.inMaintenance')}`
    : ''

  if (running || expectingTriggerLaunch) {
    return `${t('card.launchTrigger.lastSync.syncing')}${mantenanceSuffix}`
  }

  if (lastSuccessDate) {
    if (getDifferenceInMillisecondes(lastSuccessDate) < DEFAULT_TIME) {
      return `${t('card.launchTrigger.lastSync.justNow')}${mantenanceSuffix}`
    }

    return `${t('card.launchTrigger.lastSync.afterSomeTimes', {
      times: formatLocallyDistanceToNow(lastSuccessDate)
    })}${mantenanceSuffix}`
  }

  if (isDisconnected(konnector, trigger)) {
    return t('card.launchTrigger.lastSync.disconnected')
  }

  return t('card.launchTrigger.lastSync.unknown')
}
