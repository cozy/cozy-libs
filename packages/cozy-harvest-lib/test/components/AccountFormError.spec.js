import React from 'react'
import { shallow } from 'enzyme'

import { AccountFormError } from 'components/AccountForm/Error'
import { KonnectorJobError } from 'helpers/konnectors'

const fixtures = {
  konnector: {
    name: 'Konnectest',
    vendor_link: 'https://cozy.io'
  },
  error: new KonnectorJobError('LOGIN_FAILED')
}

const tMock = jest.fn().mockImplementation(key => key)

describe('AccountFormError', () => {
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {})
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  const props = {
    error: fixtures.error,
    konnector: fixtures.konnector,
    t: tMock
  }

  it('should render', () => {
    const component = shallow(<AccountFormError {...props} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render regular Error', () => {
    const component = shallow(
      <AccountFormError
        {...props}
        error={new Error('Something is undefined')}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
