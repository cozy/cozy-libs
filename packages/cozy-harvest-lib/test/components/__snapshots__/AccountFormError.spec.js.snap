// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountFormError should render 1`] = `
<div
  className="u-error"
>
  <Markdown
    source="error.job.LOGIN_FAILED.description"
  />
</div>
`;

exports[`AccountFormError should render regular Error 1`] = `
<div
  className="u-error"
>
  <Markdown
    source="error.job.UNKNOWN_ERROR.description (Something is undefined)"
  />
</div>
`;
